src/realm/query.cpp

```diff
@@ -1312,15 +1314,18 @@ void Query::do_find_all(TableView& ret, size_t limit) const
             auto best = find_best_node(pn);
             auto node = pn->m_children[best];
             if (node->has_search_index()) {
+                auto keys = node->index_based_keys();
+                REALM_ASSERT(keys);
                 KeyColumn& refs = ret.m_key_values;
 
                 // The node having the search index can be removed from the query as we know that
                 // all the objects will match this condition
                 pn->m_children[best] = pn->m_children.back();
                 pn->m_children.pop_back();
 
-                auto keys = node->index_based_keys();
-                for (auto key : keys) {
+                const size_t num_keys = keys->size();
```

Comment:
The iteration pattern here isn't really part of the optimization. The change is because  `keys` was previously a `std::vector<ObjKey>`  and now it is a `IndexEvaluator*` which doesn't have built in iterator support (I suppose I could add iterator support if you really wanted to keep the loop the same). The change in type allows us to dynamically fetch the keys directly from BPTree one at a time as needed. The previous way was to always evaluate all the results and return them in the vector. But that is entirely unused when we only have one query condition and are doing a count of results. So the optimization here is to fetch the matching object keys as needed because sometimes we don't need them at all.

Reply:
I think the iterator is a "nice to have" feature, but that's ok to leave it as it is, it is up to you.